LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+
+```solidity
+event StakingModuleShareLimitSet(
+    uint256 indexed stakingModuleId,
+    uint256 stakeShareLimit,
+    uint256 priorityExitShareThreshold,
+    address setBy
+);
+
+event StakingModuleMaxDepositsPerBlockSet(
+    uint256 indexed stakingModuleId,
+    uint256 maxDepositsPerBlock,
+    address setBy
+);
+
+event StakingModuleMinDepositBlockDistanceSet(
+    uint256 indexed stakingModuleId,
+    uint256 minDepositBlockDistance,
+    address setBy
+);
+
+error ZeroAddressStakingModule();
+error InvalidStakeShareLimit();
+error InvalidFeeSum();
+error InvalidPriorityExitShareThreshold();
+error InvalidMinDepositBlockDistance();
+
+struct StakingModule {
+    /// @notice unique id of the staking module
+    uint24 id;
+    /// @notice address of staking module
+    address stakingModuleAddress;
+    /// @notice part of the fee taken from staking rewards that goes to the staking module
+    uint16 stakingModuleFee;
+    /// @notice part of the fee taken from staking rewards that goes to the treasury
+    uint16 treasuryFee;
+    /// @notice maximum stake share that can be allocated to a module, in BP
+    uint16 stakeShareLimit; // formerly known as `targetShare`
+    /// @notice staking module status if staking module can not accept the deposits or can participate in further reward distribution
+    uint8 status;
+    /// @notice name of staking module
+    string name;
+    /// @notice block.timestamp of the last deposit of the staking module
+    /// @dev NB: lastDepositAt gets updated even if the deposit value was 0 and no actual deposit happened
+    uint64 lastDepositAt;
+    /// @notice block.number of the last deposit of the staking module
+    /// @dev NB: lastDepositBlock gets updated even if the deposit value was 0 and no actual deposit happened
+    uint256 lastDepositBlock;
+    /// @notice number of exited validators
+    uint256 exitedValidatorsCount;
+    /// @notice module's share threshold, upon crossing which, exits of validators from the module will be prioritized, in BP
+    uint16 priorityExitShareThreshold;
+    /// @notice the maximum number of validators that can be deposited in a single block
+    /// @dev must be harmonized with `OracleReportSanityChecker.churnValidatorsPerDayLimit`
+    /// (see docs for the `OracleReportSanityChecker.setChurnValidatorsPerDayLimit` function)
+    uint64 maxDepositsPerBlock;
+    /// @notice the minimum distance between deposits in blocks
+    /// @dev must be harmonized with `OracleReportSanityChecker.churnValidatorsPerDayLimit`
+    /// (see docs for the `OracleReportSanityChecker.setChurnValidatorsPerDayLimit` function)
+    uint64 minDepositBlockDistance;
+}
+
+struct StakingModuleCache {
+    address stakingModuleAddress;
+    uint24 stakingModuleId;
+    uint16 stakingModuleFee;
+    uint16 treasuryFee;
+    uint16 stakeShareLimit;
+    StakingModuleStatus status;
+    uint256 activeValidatorsCount;
+    uint256 availableValidatorsCount;
+}
```

Comment:
Consider do not pay attention to changes in internal structures

Reply:
OK, I've removed internal structures from code examples.